e2e/cephfs.go

```diff
@@ -506,6 +506,97 @@ var _ = Describe("cephfs", func() {
 				}
 			})
 
+			By("Test snapshot retention feature", func() {
+				// Delete the PVC after creating a snapshot,
+				// this should work because of the snapshot
+				// retention feature. Restore a PVC from that
+				// snapshot.
+				// snapshot beta is only supported from v1.17+
+				if k8sVersionGreaterEquals(f.ClientSet, 1, 17) {
+					err := createCephFSSnapshotClass(f)
+					if err != nil {
+						e2elog.Failf("failed to create CephFS snapshotclass with error %v", err)
+					}
+					pvc, err := loadPVC(pvcPath)
+					if err != nil {
+						e2elog.Failf("failed to load PVC with error %v", err)
+					}
+
+					pvc.Namespace = f.UniqueName
+					err = createPVCAndvalidatePV(f.ClientSet, pvc, deployTimeout)
+					if err != nil {
+						e2elog.Failf("failed to create PVC with error %v", err)
+					}
+
+					app, err := loadApp(appPath)
+					if err != nil {
+						e2elog.Failf("failed to load application with error %v", err)
+					}
+
+					app.Namespace = f.UniqueName
+					app.Spec.Volumes[0].PersistentVolumeClaim.ClaimName = pvc.Name
+					wErr := writeDataInPod(app, f)
+					if wErr != nil {
+						e2elog.Failf("failed to  write data  with error %v", wErr)
+					}
+
+					snap := getSnapshot(snapshotPath)
+					snap.Namespace = f.UniqueName
+					snap.Spec.Source.PersistentVolumeClaimName = &pvc.Name
+					// create snapshot
+					snap.Name = f.UniqueName
+					err = createSnapshot(&snap, deployTimeout)
+					if err != nil {
+						e2elog.Failf("failed to create snapshot (%s): %v", snap.Name, err)
+					}
+
+					// Delete the parent pvc before restoring
+					// another one from snapshot.
+					err = deletePVCAndApp("", f, pvc, app)
+					if err != nil {
+						e2elog.Failf("failed to delete PVC or application with error %v", err)
+					}
+
+					pvcClone, err := loadPVC(pvcClonePath)
+					if err != nil {
+						e2elog.Failf("failed to load PVC with error %v", err)
+					}
+
+					appClone, err := loadApp(appClonePath)
+					if err != nil {
+						e2elog.Failf("failed to load application with error %v", err)
+					}
+
+					pvcClone.Namespace = f.UniqueName
+					appClone.Namespace = f.UniqueName
+					pvcClone.Spec.DataSource.Name = snap.Name
+
+					// create PVC from the snapshot
+					name := f.UniqueName
+					err = createPVCAndApp(name, f, pvcClone, appClone, deployTimeout)
+					if err != nil {
+						e2elog.Logf("failed to create PVC and app (%s): %v", f.UniqueName, err)
+					}
+
+					// delete clone and app
+					err = deletePVCAndApp(name, f, pvcClone, appClone)
+					if err != nil {
+						e2elog.Failf("failed to delete PVC and app (%s): %v", f.UniqueName, err)
+					}
```

Comment:
I guess this is better for code readability, anyway its just an e2e and is passed now :)

Reply:
but increases the code duplication :), sometimes we use helper functions  to avoid code duplication

Reply:
On a second look I think validatePVCAndAppBinding() can not be used here because it takes pvcPath and appPath as argument and not pvc and app objects because of which I can not tell this function to use `DataSource.Name = snap.Name`

Let me know if I miss anything here.

Reply:
If you want I can create one more wrapper which takes the objects instead of path, or please point me to one which already takes that.

Reply:
Yes, you are correct. let's have it now later we can create the wrapper.